### ISSUE TITLE
enable no-use-before-define in eslint:recommended

### ISSUE THREAD

**Issue by aladdin-add** [eslint/eslint#11903]
refs: #11901

**Comment by g-plane** [comment]
However, use cases below are valid but reported:

```js
// functions
mayaka()

function mayaka() {}


// ES2015 classes
new Hyouka()

class Hyouka {}
```

So we may need to set option as `["error", { "functions": false, "classes": false }]` for better UX.

**Comment by aladdin-add** [comment]
I would propose the option to be default(semver-major) -- if you put it that way.

**Comment by platinumazure** [comment]
I don't think no-use-before-defined rises to the level of recommendable, in my opinion. It's partially stylistic and it's intentionally more opinionated than most of our rules. I don't feel strongly, though, and will defer to consensus.

**Comment by aladdin-add** [comment]
IMHO, since ES6, the rule is mainly to catch the usage of vars in TDZ -- it will cause a ReferenceError:

```js
function foo () {
bar;  // ReferenceError, since let/const will not be hoisted.

let bar = 0;
}
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let#Temporal_dead_zone

I agreed function & class is more like a stylistic issue. 😄

**Comment by platinumazure** [comment]
Hmm... Given the TDZ issues, I think I might agree we should do something, I'm just not sure what.

I think one option is to change the defaults of the rule to focus on variables only, which could then allow the rule to go into recommended. Another option is to deprecate this rule and create two new rules, one for variables only and one for hoisted declarations.

Neither of these feel like ideal user experiences. I think I slightly prefer changing the rule defaults.

**Comment by g-plane** [comment]
I prefer to changing default option personally, because at first time when I enabled this rule I was confused with the snippet I given above.

**Comment by mysticatea** [comment]
I think that we should move about clear TDZ errors to `no-undef` rule from `no-use-before-define` rule. TDZ is not to initialize variables with `undefined`, but it's undefined.

----

> ```js
> // ES2015 classes
> new Hyouka()
> 
> class Hyouka {}
> ```
> @g-plane 

It's TDZ error :)

**Comment by g-plane** [comment]
@mysticatea Thanks for correcting!

**Comment by dandv** [comment]
I was about to submit the same proposal, because of a costly bug that was caused by attempting to use a variable before declaring it as a `const`. At a more careful look, my IDE (WebStorm) had flagged the issue, but ESLint had not.

**Comment by platinumazure** [comment]
@mysticatea

> I think that we should move about clear TDZ errors to `no-undef` rule from `no-use-before-define` rule. TDZ is not to initialize variables with `undefined`, but it's undefined.

Do you want to create a new issue with this proposal?

**Comment by kaicataldo** [comment]
Friendly ping - how do we want to move forward here?

**Comment by btmills** [comment]
Since there's not yet a clear direction here, we decided in today's TSC meeting not to include this in the v7.0.0 release. If you'd like to keep working on this, we've moved it to the v8.0.0 project as a possible change for that release!

**Comment by btmills** [comment]
We would welcome a new issue with a concrete proposal for reporting TDZ errors if someone is willing to work on that. After discussing this as a candidate for the v8 release, we decided that we don't want to change this default, so I'll close this issue.
